File: Dynamic Programming I/P62.cpp
// Date: Mon, Sep 20, 2021
// 62. Unique Paths (https://leetcode.com/problems/unique-paths/)

// recursion approach (however it is inefficient)
// helper function
// int solve(int x1, int y1, int x2, int y2) {
// 	if (x1 == x2 && y1 == y2)
// 		return 1;
// 	if (x1 > x2 || y1 > y2)
// 		return 0;
// 	return solve(x1 + 1, y1, x2, y2) + solve(x1, y1 + 1, x2, y2);
// }

// int uniquePaths(int m, int n) {
// 	return solve(1, 1, m, n);
// }

// memoization approach
// time complexity O(m*n) and space complexity O(m*n)
// int solve(int x1, int y1, int x2, int y2, vector<vector<int>>& memo) {
//     if (x1 == x2 && y1 == y2)
//         return 1;
//     if (x1 > x2 || y1 > y2)
//         return 0;
//     if (memo[y1][x1] != -1)
//         return memo[y1][x1];
//     memo[y1][x1] = solve(x1 + 1, y1, x2, y2, memo) + solve(x1, y1 + 1, x2, y2, memo);
//     return memo[y1][x1];
// }

// int uniquePaths(int m, int n) {
//     vector<vector<int>> memo(m + 1, vector<int>(n + 1, -1));
//     return solve(1, 1, n, m, memo);
// }

// TODO: practice using dynamic programming approach
// dp[y1][x1] = dp[y1 - 1][x1] + dp[y1][x1 - 1]
// Time and space: O(mn)
int uniquePaths(int m, int n) {
    vector<vector<int>> dp(m + 1, vector<int>(n + 1));
    for (int i = 1; i <= m; i++) {
        for (int j = 1; j <= n; j++) {
            if (i == 1 || j == 1) 
                dp[i][j] = 1;
            else
                dp[i][j] = dp[i-1][j] + dp[i][j-1];
        }
    }
    return dp[m][n];
}

File: Algorithm-II/6-Dynamic-Programming/P62.java
package algorithm2.dynamicProgramming;

/**
 * Date: December 27, 2021 - Redo: August 01, 2022 (Leetcode Daily Challenge)
 */

public class P62 {
    /*  recursion approach (Time Limit Exceed)
        public int uniquePaths(int m, int n) {
            return dfs(0, 0, m, n);
        }

        private int dfs(int row, int col, int m, int n) {
            int count = 0;
            if (row == m - 1 && col == n - 1)
                return 1;
            if (row < m - 1) count += dfs(row + 1, col, m, n);
            if (col < n - 1) count += dfs(row, col + 1, m, n);
            return count;
        }
     */

    // dynamic programming -> solve overlapped sub-problems exist in recursion approach
    // time: O(m*n)
    // space: O(m*n)
    public int uniquePaths(int m, int n) {
        int[][] dp = new int[m][n];
        for (int i = 0; i < m; i++) {
            for (int j = 0; j < n; j++) {
                if (i == 0 || j == 0) dp[i][j] = 1;
                else dp[i][j] = dp[i - 1][j] + dp[i][j - 1];
            }
        }
        return dp[m - 1][n - 1];
    }

    public static void main(String[] args) {
        P62 p = new P62();
        System.out.println("-----------------------------------");
        System.out.println("Input: m = 3, n = 7 -> Expect: 28");
        System.out.println("Output: " + p.uniquePaths(3, 7));
        System.out.println("-----------------------------------");
        System.out.println("Input: m = 3, n = 2 -> Expect: 3");
        System.out.println("Output: " + p.uniquePaths(3, 2));
        System.out.println("-----------------------------------");
        System.out.println("Input: m = 51, n = 9 -> Expect: 1916797311");
        System.out.println("Output: " + p.uniquePaths(51, 9));
    }
}